### PR TITLE
Domain services

### PR DESCRIPTION
**Pull request by ivanpaulovich**
* Designed 3 domain services
* Customer aggregate requires `AccountId` instead of `IAccount`
* Account aggregate requires `CustomerId` instead of `ICustomer`
* Aggregates and Entities set to `abstract`

### CODE REVIEW

File: src/Application/UseCases/Deposit.cs [hunk 1]
```diff
@@ -4,47 +4,45 @@ namespace Application.UseCases
...
-            IAccount account;
-
             try
             {
-                account = await _accountRepository.Get(input.AccountId);
+                var account = await _accountRepository.Get(input.AccountId);
+
+                var credit = await _accountService.Deposit(account, input.Amount);
+
+                await _accountRepository.Update(account, credit);
```
Comment (bommen) [inline src/Application/UseCases/Deposit.cs]: The repository Update method is also called from within the service. Isn't it a cleaner approach to leave it there and omit the repository call from the Use Cases?

File: src/Domain/Accounts/AccountService.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            _accountFactory = accountFactory;
+            _accountRepository = accountRepository;
+        }
+
+        public async Task<IAccount> OpenCheckingAccount(CustomerId customerId, PositiveMoney amount)
+        {
+            var account = _accountFactory.NewAccount(customerId);
+            var credit = account.Deposit(_accountFactory, amount);
+            await _accountRepository.Add(account, credit);
```
Comment (bommen) [inline src/Domain/Accounts/AccountService.cs]: Method already invoked by AccountService.Deposit

File: src/Domain/Accounts/Account.cs [hunk 3]
```diff
@@ -2,10 +2,10 @@ namespace Domain.Accounts
...
     using System;
     using Domain.Accounts.Credits;
-    using Domain.Accounts.Debits;
-    using Domain.Accounts.ValueObjects;
-
-    public class Account : IAccount
+    using Domain.Accounts.Debits;
+    using Domain.Accounts.ValueObjects;
+
+    public abstract class Account : IAccount
```
Comment (bommen) [inline src/Domain/Accounts/Account.cs]: What is the reasoning behind making the domain classes abstract?

File: src/Domain/Accounts/Credits/Credit.cs [hunk 4]
```diff
@@ -1,9 +1,9 @@
 namespace Domain.Accounts.Credits
 {
-    using System;
-    using Domain.Accounts.ValueObjects;
-
-    public class Credit : ICredit
+    using System;
+    using Domain.Accounts.ValueObjects;
+
+    public abstract class Credit : ICredit
```
Comment (bommen) [inline src/Domain/Accounts/Credits/Credit.cs]: What is the reasoning behind making the domain classes abstract? (Same as Account)

File: src/Domain/Accounts/Debits/Debit.cs [hunk 5]
```diff
@@ -1,9 +1,9 @@
 namespace Domain.Accounts.Debits
 {
-    using System;
-    using Domain.Accounts.ValueObjects;
-
-    public class Debit : IDebit
+    using System;
+    using Domain.Accounts.ValueObjects;
+
+    public abstract class Debit : IDebit
```
Comment (bommen) [inline src/Domain/Accounts/Debits/Debit.cs]: What is the reasoning behind making the domain classes abstract? (See above)

File: src/Domain/Customers/Customer.cs [hunk 6]
```diff
@@ -1,9 +1,9 @@
 namespace Domain.Customers
 {
-    using Domain.Accounts;
-    using Domain.Customers.ValueObjects;
-
-    public class Customer : ICustomer
+    using Domain.Accounts.ValueObjects;
+    using Domain.Customers.ValueObjects;
+
+    public abstract class Customer : ICustomer
```
Comment (bommen) [inline src/Domain/Customers/Customer.cs]: What is the reasoning behind making the domain classes abstract? (See above)

File: src/Domain/Security/User.cs [hunk 7]
```diff
@@ -1,10 +1,9 @@
...
 {
-    using Domain.Customers;
-    using Domain.Customers.ValueObjects;
-    using Domain.Security.ValueObjects;
-
-    public class User : IUser
+    using Domain.Customers.ValueObjects;
+    using Domain.Security.ValueObjects;
+
+    public abstract class User : IUser
```
Comment (bommen) [inline src/Domain/Security/User.cs]: What is the reasoning behind making the domain classes abstract? (See above)

**Comment by ivanpaulovich** [comment]
Just to clarify some patterns:

* the intent with DomainServices is to create a reusable business rules against aggregates and repositories, they should not invoke `UnitOfWork.Save()`.
* the use cases are the coordination object and they should invoke `UnitOfWork.Save()` in case of writing operations.

**Pull request merged**